gnome-initial-setup/gis-page-util.c

```diff
@@ -492,3 +491,94 @@ gis_page_util_show_demo_dialog (GisPage *page)
   gtk_window_set_modal (GTK_WINDOW (demo_dialog), TRUE);
   gtk_window_present (GTK_WINDOW (demo_dialog));
 }
+
+static gboolean
+run_dialog_cb (gpointer user_data)
+{
+  GtkDialog *dialog = GTK_DIALOG (user_data);
+  gtk_dialog_run (dialog);
+  return G_SOURCE_REMOVE;
+}
+
+static void
+on_reformatter_exited (GTask  *task,
+                       GError *error)
+{
+  GisPage *page = GIS_PAGE (g_task_get_source_object (task));
+  GisDriver *driver = page->driver;
+
+  gis_driver_show_window (driver);
+
+  if (error == NULL)
+    {
+      g_task_return_boolean (task, TRUE);
+    }
+  else
+    {
+      GtkWindow *toplevel = GTK_WINDOW (gtk_widget_get_toplevel (GTK_WIDGET (page)));
+      GtkWidget *message_dialog;
+
+      g_critical ("Error running the reformatter: %s", error->message);
+
+      message_dialog = gtk_message_dialog_new (toplevel,
+                                               GTK_DIALOG_MODAL | GTK_DIALOG_DESTROY_WITH_PARENT,
+                                               GTK_MESSAGE_ERROR,
+                                               GTK_BUTTONS_CLOSE,
+                                               _("Error running the reformatter: %s"), error->message);
+      g_idle_add_full (G_PRIORITY_DEFAULT_IDLE, run_dialog_cb,
+                       message_dialog, (GDestroyNotify) gtk_widget_destroy);
+
+      g_task_return_error (task, g_steal_pointer (&error));
+    }
+}
+
+static void
+reformatter_exited_cb (GObject      *source,
+                       GAsyncResult *result,
+                       gpointer      user_data)
+{
+  g_autoptr(GTask) task = G_TASK (user_data);
+  g_autoptr(GError) error = NULL;
+
+  g_subprocess_wait_check_finish (G_SUBPROCESS (source), result, &error);
+  on_reformatter_exited (task, g_steal_pointer (&error));
+}
+
+/**
+ * gis_page_util_run_reformatter:
+ *
+ * Launches the reformatter, and arranges for the assistant to be hidden for
+ * the duration of its runtime. @callback will be called when the reformatter
+ * exits.
+ *
+ * There is no corresponding _finish() function because (at present) neither
+ * caller actually cares about the result.
+ */
+void
+gis_page_util_run_reformatter (GisPage            *page,
+                               GAsyncReadyCallback callback,
+                               gpointer            user_data)
+{
+  g_autoptr(GTask) task = g_task_new (page, NULL, callback, user_data);
```

Comment:
I don't understand why a `GTask` is needed here. AFAICS you are only using it to pass some info around, and never actually run any task with it, or did I miss something?

Reply:
Also, if you didn't need the `GTask`, the `on_reformatter_exited` code could be simplified and its use made more consistent: sometimes it's verifying the error before calling it, others just calling it directly.

Reply:
I am using the GTask to:

a) represent an ongoing async operation
b) store a reference to the page, a callback, and some user data
c) call that callback when the operation is complete

It's true that there's no _finish() function but I don't think writing my own struct to hold the callback and call it would be a net win.

Reply:
I don't think the `on_reformatter_exited` code can be made simpler because it is called in two cases:

1. spawning the task failed entirely
2. having been successfully spawned, the task exited some time later, and we check its exit status

Reply:
I don't know. It feels a bit weird to me that you're essentially using a `GTask` as a struct for storing a page and a callback. So IMO a struct with those two members would just be simpler (reading the code I kept expecting to see something running the `GTask`).

But I won't block this if you feel confident that this is a good idea...